### PR TITLE
test: add type hints and `mypy` linting

### LINKED ISSUE #232: Annotate code with type hints
Type hints help catch certain errors, document the code base, help static analyzers (like [mypy](http://mypy-lang.org/) or Google's [pytype](https://github.com/google/pytype)) and linters in IDEs, and let us build and maintain a cleaner architecture overall.

Annotations can be done iteratively over multiple, small Pull Requests, based on the docstrings already in place. Type hints can be tested with local installations of mypi or pytype until these static analyzers are integrated in the CI pipelines.

See https://docs.python.org/3/library/typing.html for more details.

### PR DESCRIPTION
**Pull request by faissaloux**
Hello 👋🏻 

On this PR I have added type hints ✨ 
resolves #232

### CODE REVIEW

**Comment by lvaylet** [comment]
Hi @faissaloux and thanks a lot for your contribution. I like type hints too. They make Python feel a little less interpreted and a little more statically typed. I will definitely take a look at this PR once I am done with adding the Cloud Monitoring MQL backend. Some commits I made there will make our life easier when it comes to linting and testing.

File: slo_generator/backends/cloud_service_monitoring.py
Comment (lvaylet) [inline slo_generator/backends/cloud_service_monitoring.py:360]: Move the `# pylint: disable` directive to the line above and `make lint` should succeed.

I just discovered that directives such as this one must be on the same line as `def`.

File: mypy.ini
Comment (lvaylet) [inline mypy.ini:3]: I understand the value of `ignore_missing_imports`. What kind of errors/warnings are we ignoring with ` disable_error_code`? Do these warnings/errors prevent the MyPy validation from completing successfully?

Also add a newline at the end of the file.

**Review by lvaylet — CHANGES_REQUESTED** [review]
Thanks a lot for these annotations and the new MyPy target in the Makefile. I added a couple of suggestions for improvement.

File: slo_generator/backends/cloud_service_monitoring.py
Comment (faissaloux) [reply slo_generator/backends/cloud_service_monitoring.py]: I have added those options to `.pylintrc`.

File: mypy.ini
Comment (faissaloux) [reply mypy.ini:3]: Yes MyPy fails without that ignoring

image

File: .pylintrc
Comment (lvaylet) [inline .pylintrc:69]: Not sure about this one, for two reasons:
- The change is not related to the task at hand in this PR, which is to add type hints to the existing code.
- We do not necessarily want to disable these warnings **globally**. I agree it makes the code cleaner, with fewer annotations. On the other hand, these warnings are valuable and we only want to disable them for a few occurrences **locally**.

File: slo_generator/api/main.py
Comment (lvaylet) [inline slo_generator/api/main.py:194]: For the reasons discussed above, I'd rather keep this annotation. It indicates something might be wrong and we could address it properly in a different PR.

File: slo_generator/backends/cloud_service_monitoring.py
Comment (lvaylet) [inline slo_generator/backends/cloud_service_monitoring.py:360]: Same here. Just move the annotation to line 359 so the linter takes it into account.

File: slo_generator/exporters/cloudevent.py
Comment (lvaylet) [inline slo_generator/exporters/cloudevent.py:29]: Same here. We do not want to disable this warning **globally**.

**Review by lvaylet — CHANGES_REQUESTED** [review]

File: .pylintrc
Comment (lvaylet) [inline .pylintrc:69]: Something unexpected happened here. You probably kept a line that was supposed to be removed, and vice versa. As a result, we are getting a lot of warnings during `make lint`: https://github.com/google/slo-generator/actions/runs/3204729299/jobs/5236369774. You probably just need to restore the original value.

File: slo_generator/backends/cloud_service_monitoring.py
Comment (lvaylet) [inline slo_generator/backends/cloud_service_monitoring.py:359]: I just noticed that the `delete()` function above (line 122) has a similar multi-line signature with a Pylint annotation. It is probably cleaner to put the annotation on a separate line, just above the function name (assuming it still works as expected).

Apologies for the back and forth on this one, but I find it cleaner this way.

**Review by lvaylet — CHANGES_REQUESTED** [review]

File: .pylintrc
Comment (faissaloux) [reply .pylintrc:69]: Yeah I need to keep the `logging-fstring-interpolation` you said before I have to remove it because it's not related to this PR.

Should I keep it now! 👀

File: slo_generator/backends/cloud_service_monitoring.py
Comment (faissaloux) [reply slo_generator/backends/cloud_service_monitoring.py:359]: Yeah much cleaner. No worries I'll fix that

File: .pylintrc
Comment (faissaloux) [reply .pylintrc:69]: Yeah I think I didn't understood your first change.

I'll fix it

File: mypy.ini
Comment (lvaylet) [reply mypy.ini]: I'd rather not disable these warnings globally. They probably can be adressed locally. If not, let's check if they can be disabled locally too.

For example, I am pretty sure the `[misc]` errors can easily be fixed by reordering the variable declarations in `report.py`. PyCharm (the IDE I am using) highlights these errors and makes it easy to confirm everything is in the right order without having to run `mypy` again.

File: Makefile
Comment (lvaylet) [inline Makefile:20]: `mypy` is only called once so we don't need this variable. Just call it directly in the target.

File: Makefile
Comment (lvaylet) [inline Makefile:77]: I just noticed this target. Can you rename it to `mypy` to match the one with `pytype` below? Then move it just below the `pytype`target and call it from the `lint` target on line 75 so it runs automatically on every `make lint` call. Like `lint: flake8 pylint pytype mypy`.

File: mypy.ini
Comment (lvaylet) [inline mypy.ini:3]: Let's work on fixing as many warnings/errors as possible in these categories instead of disabling the errors globally. The `misc` category should be easy. It is just a matter of rearranging the variable declarations. A good IDE like PyCharm can probably do this automatically (haven't tested it myself).

File: slo_generator/backends/prometheus.py
Comment (lvaylet) [inline slo_generator/backends/prometheus.py:106]: Are you sure about the `tuple[float, float]` return type?

File: slo_generator/backends/prometheus.py
Comment (lvaylet) [inline slo_generator/backends/prometheus.py:167]: Are you sure about the `float` return type?

File: slo_generator/backends/prometheus.py
Comment (lvaylet) [inline slo_generator/backends/prometheus.py:187]: Can we specialize `list` here to indicate which type is in the list?

File: slo_generator/backends/prometheus.py
Comment (lvaylet) [inline slo_generator/backends/prometheus.py:188]: Same here with `dict`.

File: slo_generator/compute.py
Comment (lvaylet) [inline slo_generator/compute.py:34]: Not sure we need the `Optional` there. The default value already indicates that the value is optional. From my experience, we mostly use `Optional` on return types but I have done an extensive research on this topic. Are you aware of any best practice or article we could use as a reference?

File: slo_generator/compute.py
Comment (lvaylet) [inline slo_generator/compute.py:36]: Same here.

File: slo_generator/compute.py
Comment (lvaylet) [inline slo_generator/compute.py:37]: Same here.

File: slo_generator/compute.py
Comment (lvaylet) [inline slo_generator/compute.py:99]: Same here.

File: slo_generator/migrations/migrator.py
Comment (lvaylet) [inline slo_generator/migrations/migrator.py:51]: Same here.

File: slo_generator/migrations/migrator.py
Comment (lvaylet) [inline slo_generator/migrations/migrator.py:52]: Same here.

File: slo_generator/migrations/migrator.py
Comment (lvaylet) [inline slo_generator/migrations/migrator.py:217]: Same here.

File: slo_generator/migrations/migrator.py
Comment (lvaylet) [inline slo_generator/migrations/migrator.py:252]: Same here.

File: slo_generator/migrations/migrator.py
Comment (lvaylet) [inline slo_generator/migrations/migrator.py:293]: Same here.

File: slo_generator/migrations/migrator.py
Comment (lvaylet) [inline slo_generator/migrations/migrator.py:441]: Same here.

File: slo_generator/migrations/migrator.py
Comment (lvaylet) [inline slo_generator/migrations/migrator.py:558]: Same here.

File: slo_generator/utils.py
Comment (lvaylet) [inline slo_generator/utils.py:50]: Same here.

File: slo_generator/utils.py
Comment (lvaylet) [inline slo_generator/utils.py:70]: Same here.

File: slo_generator/utils.py
Comment (lvaylet) [inline slo_generator/utils.py:111]: Same here.

File: slo_generator/utils.py
Comment (lvaylet) [inline slo_generator/utils.py:191]: Same here.

File: slo_generator/utils.py
Comment (lvaylet) [inline slo_generator/utils.py:344]: Same here.

File: slo_generator/utils.py
Comment (lvaylet) [inline slo_generator/utils.py:504]: Same here.

**Review by lvaylet — CHANGES_REQUESTED** [review]
A lot of comments but not so many changes to make at the end of they day :-)

**Comment by faissaloux** [comment]
@lvaylet heeey sorry for late response 😅 I've got busy
I just found some time so I can check your comments now 😃

File: mypy.ini
Comment (faissaloux) [reply mypy.ini]: `misc` errors are fixed

File: Makefile
Comment (faissaloux) [reply Makefile]: Done

File: Makefile
Comment (faissaloux) [reply Makefile]: Done

File: mypy.ini
Comment (faissaloux) [reply mypy.ini]: Fixed all error except the `ignore_missing_imports` which I have no control to fix, since the third party packages that not supporting typing are the cause of this.

File: slo_generator/backends/prometheus.py
Comment (faissaloux) [reply slo_generator/backends/prometheus.py:167]: Yeah sure check line 177

File: slo_generator/backends/prometheus.py
Comment (faissaloux) [reply slo_generator/backends/prometheus.py:106]: Yes sure.

The `good_count` is coming from `count()` which returns float [line 177]
And the `bad_count` is coming from a sub of two floats coming of from `count()` which returns float [line 177]

File: slo_generator/backends/prometheus.py
Comment (faissaloux) [reply slo_generator/backends/prometheus.py:187]: Yes it's a `list[str]` 😄
Done 🚀

File: slo_generator/compute.py
Comment (faissaloux) [reply slo_generator/compute.py]: Yeah you right, I got some errors about this in first time I thought it's required, but I think it's not 😅 I have fixed this

File: slo_generator/backends/prometheus.py
Comment (lvaylet) [reply slo_generator/backends/prometheus.py:167]: Agreed, but then it makes little sense to call the function `count`, document the return type as `int` in the docstring and return `NO_DATA` when an exception is raised. I am pretty sure we should replace `float` with `int` everywhere. Let me ask Olivier who wrote this code in the first place.

File: slo_generator/exporters/prometheus_self.py
Comment (lvaylet) [inline slo_generator/exporters/prometheus_self.py:32]: `dict` (with a lowercase `d` at the beginning) is probably enough here as we do not specify the key/value types. This works in Python 3.6 and 3.7 too (as we need to support these versions). Then you can remove the `typing` import on line 19.

File: slo_generator/report.py
Comment (lvaylet) [inline slo_generator/report.py:50]: Add an empty line and a `# SLI` comment above this line to clearly indicate this variable is related to an SLI. Otherwise we are under the impression it belongs to the SLO section now that the variable definitions have moved elsewhere.

File: slo_generator/report.py
Comment (lvaylet) [inline slo_generator/report.py:73]: Same here. Add an empty line and a `# SLO` comment above this line.

File: slo_generator/report.py
Comment (lvaylet) [inline slo_generator/report.py:262]: `dict` is probably enough here. Then you can remove the `Dict` and `Any` imports on line 21.

**Review by lvaylet — CHANGES_REQUESTED** [review]
Looking much better now! Just a few more changes and we should be good to go.

Also check the output of the CI for a handful of changes to make on `Optional[str]` arguments: https://github.com/google/slo-generator/actions/runs/3259133646

Finally, no need to apologize for being busy. This is an open source project and I am really happy you take some personal time to work on it. No pressure.

File: slo_generator/backends/prometheus.py
Comment (faissaloux) [reply slo_generator/backends/prometheus.py:167]: Yeah right I got confused the first time checking this function, the docstring says it returns `int` but on return `line 177` there's `float`.

I'll wait Olivier's response 👀

**Comment by faissaloux** [comment]
About the `Optional` arguments, I think we should bring back the `Optional` type so the lint tests can pass.

What do you think!

File: .gitignore
Comment (lvaylet) [inline .gitignore:49]: Add a newline at the end of the file so GitHub stops complaining?

File: Makefile
Comment (lvaylet) [inline Makefile:85]: Not sure we need to perform a `clean` before calling `mypy` here.

File: Makefile
Comment (lvaylet) [inline Makefile:86]: Not sure we need to call `python` here. Call `mypy` directly, like `pytype` above.

**Review by lvaylet — CHANGES_REQUESTED** [review]
Thanks a lot. I like how this looks.

You are right about bringing back `Optional` for the checks.

File: slo_generator/compute.py
Comment (lvaylet) [reply slo_generator/compute.py:35]: https://docs.python.org/3/library/typing.html#typing.Optional says it is OK to use `Optional` when `None` is a valid value.

**Review by lvaylet — CHANGES_REQUESTED** [review]
Just a couple small changes left.

Also, please stick to the original code as much as possible. I do not fully understand the purpose of the new code you introduced in `main.py` or other files (though I am sure there is a purpose :-)). Let's focus on type hints in this PR.

**Comment by faissaloux** [comment]
Oow yeah you right I wanted to but I found my self changing it because lint keeps failing so I have to fix it 👀

About `main.py` I've got `None has no attribute get` so I made sure `config` is not None before getting anything from it 

Now we still have some errors because of some duplicated code 😐

**Comment by faissaloux** [comment]
Oow yeah you right I wanted to but I found my self changing it because lint keeps failing so I have to fix it 👀

About `main.py` I've got `None has no attribute get` so I made sure `config` is not None before getting anything from it .. or should I reverse it despite of the failing lint!!

Now we still have some errors because of some duplicated code 😐

File: slo_generator/api/main.py
Comment (faissaloux) [reply slo_generator/api/main.py:101]: Yeah I agree with you, I have modified it so the lint can pass 👀 .

Should I reverse it even the lint fails! What do you think!

File: slo_generator/backends/cloud_monitoring_mql.py
Comment (faissaloux) [reply slo_generator/backends/cloud_monitoring_mql.py:208]: Yeah you right I agree, I have done this so the lint can pass

File: slo_generator/api/main.py
Comment (faissaloux) [reply slo_generator/api/main.py:196]: Same, I'll wait your answer about this, if you agree on lint failing for now so I'll reverse it

File: slo_generator/backends/cloud_monitoring_mql.py
Comment (faissaloux) [reply slo_generator/backends/cloud_monitoring_mql.py:49]: Same, I'll wait your answer about this, if you agree on lint failing for now so I'll reverse it

File: slo_generator/backends/prometheus.py
Comment (faissaloux) [reply slo_generator/backends/prometheus.py:106]: Yeah right, but it's returning float 🤷‍♂️

Still waiting for Olivier's response 👀

File: slo_generator/api/main.py
Comment (ocervell) [reply slo_generator/api/main.py:101]: To be honest, this rule that states that mutable defaults are bad is strange to me when we never modify that mutable parameter. Maybe we can ignore it in linting, I'll let @Ivaylet be the judge ;)

File: slo_generator/api/main.py
Comment (ocervell) [reply slo_generator/api/main.py:196]: ditto

File: slo_generator/backends/cloud_monitoring_mql.py
Comment (ocervell) [reply slo_generator/backends/cloud_monitoring_mql.py:49]: this might be a bit different than the others - the client is needed in all cases, so either it is passed to mutualize clients in the main process, or created here so this is not shocking to me. In most other providers we do something like:
```
self.client = client
if not self.client:
  self.client = QueryServiceClient()
```

Comment (ocervell) [reply slo_generator/backends/cloud_monitoring_mql.py:49]: Yep, I prefer the explicit `if client is None` here for readability.

File: slo_generator/backends/prometheus.py
Comment (ocervell) [reply slo_generator/backends/prometheus.py:106]: Mmmh, need to look over the Prometheus response in that case - in some bakends good / bad are `float` because it's an average over a time window, I'm afraid turning this to `int` will decrease the overall precision of the resulting SLI...

File: slo_generator/backends/prometheus.py
Comment (ocervell) [reply slo_generator/backends/prometheus.py:167]: same as above, we need to observe some Prometheus responses and use it as-is I believe.

**Review by ocervell — CHANGES_REQUESTED** [review]
Looks good overall, thank you so much for taking the time to do this ! It will improve the dev experience and possibly ease-up some testing as well.

Some changes required, after that LGTM :)

**Comment by faissaloux** [comment]
@lvaylet @ocervell I have tried to send review request for both of you but when I send it to one it removes from the other 😅 IDK why!

Anyway have a good night 🌙

File: slo_generator/backends/cloud_monitoring_mql.py
Comment (lvaylet) [reply slo_generator/backends/cloud_monitoring_mql.py:49]: What does the linter warn us about here? Let's focus on type hints only in this PR. We can disable pylint warnings locally if it makes sense.

File: slo_generator/backends/prometheus.py
Comment (lvaylet) [reply slo_generator/backends/prometheus.py:106]: Let's keep `float` here. I will open an issue to confirm the actual type later on.

File: slo_generator/backends/prometheus.py
Comment (lvaylet) [reply slo_generator/backends/prometheus.py:167]: Same here. Let's keep `float` until we know better.

**Review by lvaylet — CHANGES_REQUESTED** [review]
Almost there! A single change left.

File: slo_generator/backends/cloud_monitoring_mql.py
Comment (faissaloux) [reply slo_generator/backends/cloud_monitoring_mql.py:49]: In here I have seen that `self.client` is assigned two times so I refactored it 👀

Comment (lvaylet) [reply slo_generator/backends/cloud_monitoring_mql.py:49]: It is not an issue here. Testing for `None` and reassigning is a common pattern in Python. Please restore the initial code.

**Review by lvaylet — CHANGES_REQUESTED** [review]
Thanks Faissal. Just restore the original code for `self.client` in `cloud_monitoring_mql.py` and we should be good to go.

Regarding the linting checks that fail, you can add `# pylint: disable=duplicate-code` comments just above the definitions of `good_bad_ratio()` and `distribution_cut()` in `cloud_monitoring.py`. Then you can check that everything works from your local machine by running `make lint`. No need to wait for the CI pipeline to trigger.

File: slo_generator/utils.py
Comment (lvaylet) [inline slo_generator/utils.py:68]: `Optional[dict]` generates an error when `pytype` does its own linting and type checking (check [the logs](https://github.com/google/slo-generator/actions/runs/3303042130/jobs/5450547176) for more details):

```
File "/home/runner/work/slo-generator/slo-generator/slo_generator/api/main.py", line 101, in run_export: No attribute 'get' on None [attribute-error]
  In Optional[dict]
File "/home/runner/work/slo-generator/slo-generator/slo_generator/api/main.py", line 195, in process_batch_req: No attribute 'get' on None [attribute-error]
  In Optional[dict]
```

Replace with `dict`? It is probably the right type here, even if there is a possiblity to raise an exception.

**Review by lvaylet — CHANGES_REQUESTED** [review]
Good job. Now we need to work on making the `pytype` tests pass. Proceed incrementally, starting with the error in `api/main.py` mentioned below in a comment. Then make sure to run `make lint` on your local machine before pushing. Ideally this is the last review. I wll proceed with merging this PR if all checks succeed.

File: slo_generator/utils.py
Comment (faissaloux) [reply slo_generator/utils.py:68]: I don't think so, if we replace it with `dict` it will complains.

We need to check if it's not null before `get` anything from it, that's what I did before

Comment (lvaylet) [reply slo_generator/utils.py:68]: You are absolutely right, and that is why I like type hints! `mypy` and `pytype` are able to surface bugs like this one. I cloned your branch and ran a few tests. `config` can indeed be `None` and must be checked explicitly. Good catch!

For the sake of brevity and simplicity, can you surround the two faulty lines (101 and 195) with a `FIXME` (so we have a chance to fix this later) and temporarily disable the `pytype` warnings? I used this on lines 101 and 195 and was able to run `make pytype` successfully as a consequence:

```python
            # pytype: disable=attribute-error
            # FIXME `load_config()` returns `Optional[dict]` so `config` can be `None`
            exporter_conf = config.get('pubsub_batch_handler')
            # pytype: enable=attribute-error
```

Do this on both lines and we should be good to go.

Comment (faissaloux) [reply slo_generator/utils.py:68]: I did this but now `pylint` fails because of `FIXME`

**Review by lvaylet — APPROVED** [review]
LGTM, thanks a lot @faissaloux!

**Pull request merged**
